packages/vite/src/node/server/index.ts

```diff
@@ -541,17 +541,22 @@ export async function createServer(
     }
   }
 
+  let isOptimized = false
   if (!middlewareMode && httpServer) {
     // overwrite listen to run optimizer before server start
     const listen = httpServer.listen.bind(httpServer)
     httpServer.listen = (async (port: number, ...args: any[]) => {
+      if (isOptimized) {
+        return listen(port, ...args)
+      }
       try {
         await container.buildStart({})
         await runOptimize()
       } catch (e) {
         httpServer.emit('error', e)
         return
       }
+      isOptimized = true
       return listen(port, ...args)
     }) as any
```

Comment:
What if we restore the original `listen` instead of having to add the extra flag? 
```suggestion
  if (!middlewareMode && httpServer) {
    // overwrite listen to run optimizer before server start
    const listen = httpServer.listen.bind(httpServer)
    httpServer.listen = (async (port: number, ...args: any[]) => {
      try {
        await container.buildStart({})
        await runOptimize()
        // successful server start, restore listen to avoid re-optimizing
        // in next call (as it happen when port is taken) #3640
        httpServer.listen = listen
      } catch (e) {
        httpServer.emit('error', e)
        return
      }
      return listen(port, ...args)
    }) as any
```

Reply:
Ignore this one, see https://github.com/vitejs/vite/pull/3640#discussion_r645977366